app/master/cluster_master.py

```diff
@@ -62,12 +65,24 @@ def api_representation(self):
             'slaves': slaves_representation,
         }
 
-    def builds(self):
+    def builds(self, offset: int=None, limit: int=None) -> List['Build']:
         """
-        Returns a list of all builds
-        :rtype: list[Build]
+        Returns a list of all builds. Both offset and limit are assumed to be nonnegative at this point
+        since the corrections or negative offsets and limits are done when the query parameters are first initialized.
```

Comment:
I'm fine with leaving negative offsets for a later change, but I just spent a long time thinking about negative offsets though and the exact spec I want, so i'm gonna write it down here. I don't really care if we implement it now, later or never. Fwiw @josephharrington, I do think we can make negative limits a well defined and intuitive concept. Maybe it's not useful, but I think it makes sense.

The mental model I have is this. Lets say you have an array of `N` builds. You can think of any `offset`/`limit` combo as defining a "window" that you place over the array of builds. The return value is just whatever part of the array you can see through that window. If `offset` is negative you mod it by `N`, if it's positive you just leave it alone.

```python
offset = offset if offset > 0 else offset % N  # Don't mod by N if offset is positive so we can handle cases where offset is off the right edge of the array
if limit >= 0:
    start, end = offset, min(offset + limit, N)
else:
    start, end = max(offset + limit, 0), min(offset, N)
return builds[start:end]
```

Some cases to test. (`N` is 100 for all cases). I do agree that the last 2 cases are a little weird, but they're well defined and they fit my model.

| offset        | limit           | return value  | case
|:------------- |:-------------| :-----| :-----|
| 5/-95 | 20      |    [5,24] | positive limit
| -5/95 | -20 | [75,95] | negative limit
| -5/95      | 20 | [95, 99] | limit overflow right
| 5/-95 | -20 | [0,4] | limit overflow left
| 100| 1 | [] | offset off right edge, positive limit
| 101| -1 | [] | offset off right edge. Window does not overlap array
| 100| -2 | [99] | offset off right edge. Window does overlap array